transform-plugins/src/main/java/co/cask/hydrator/plugin/StreamFormatter.java

```diff
@@ -0,0 +1,227 @@
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.TransformContext;
+import co.cask.cdap.etl.common.StructuredRecordStringConverter;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms a {@link StructuredRecord} into {@link co.cask.cdap.api.flow.flowlet.StreamEvent}.
+ * 
+ * This transforms formats the structured into a StreamEvent. It allows one to specify the 
+ * fields that need to be considered as HEADER and for BODY one can specify the format as CSV, 
+ * TSV, PSV and JSON.  
+ */
+@Plugin(type = "transform")
+@Name("StreamFormatter")
+@Description("Formats the data from Structured Record to CDAP Stream format.")
+public final class StreamFormatter extends Transform<StructuredRecord, StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamFormatter.class);
+  private final Config config;
+  
+  // Specifies the output schema. 
+  private Schema outSchema;
+  
+  // Fields that need to be added StreamEvent Header.
+  private String[] headerFields = null;
+  
+  // Fields that need to be included in the body. 
+  private String[] bodyFields = null;
+  
+  // Output Schema Field name that is considered as Stream Event Header. 
+  private String headerFieldName;
+  
+  // Output Schema Field name that is considered as Stream Event Body.
+  private String bodyFieldName;
+
+  // Only used in tests. 
+  public StreamFormatter(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+
+    // If the format specified is not of one of the allowed types, then throw an exception.
+    if(!config.format.equalsIgnoreCase("CSV") && !config.format.equalsIgnoreCase("TSV")
+      && !config.format.equalsIgnoreCase("JSON") && !config.format.equalsIgnoreCase("PSV")) {
+      throw new IllegalArgumentException("Invalid format '" + config.format + "', specified. Allowed values are " +
+                                           "CSV, TSV, PSV or JSON.");
+    }
+
+    // Check if the output schema JSON is invalid
+    try {
+      Schema out = Schema.parseJson(config.schema);
+      List<Schema.Field> fields = out.getFields();
+
+      if (fields.size() != 2) {
+        throw new IllegalArgumentException("Output schema should have only two fields. One of type " +
+                                             "String for Stream body and other of type Map<String, String> for " +
+                                             "Stream header.");
+      }
+
+      // We have to make sure that output schema includes two fields and only two fields. 
```

Comment:
The check below will check if each field is either a map or a string.The check below doesn't catch if field[0] is String and field[1] is also a string. Same case with Map.